### PR TITLE
[improve][doc] Add a limitation for key_shared subscription type

### LINKED ISSUE #15705: New KeyShared consumers will not get any messages until a consumer that did get messages disconnects or acks/nacks some messages
**Describe the bug**

If a KeyShared consumer has any unacknowledged messages, new KeyShared consumers on the same subscription after that point will not get any new messages (even messages with brand new keys) until the original consumer either disconnects or acks/nacks some indeterminate number of messages.

This is a really bad bug!

**Bug scenario 1**

1. Create a **non partitioned** topic.
2. Create a KeyShared subscription on that topic.
3. Produce some number of messages on the topic with a given key, say `key1`, using a KeyBased batching strategy.
4. Start a consumer on the topic with the below code and ensure it prints `Press a key to acknowledge messages`.
5. Start a second consumer on the same topic; it may or may not print `Press a key to acknowledge messages` depending on whether `key1` is assigned to it (but it will never print that if this bug is in effect).
6. Produce some number (10 should be sufficient) of messages on the topic with unique keys not equal to `key1`; say `key2`, `key3`, and so on. The goal here is to get a key that hashes to the second consumer's range.
7. Observe that the second consumer never gets a message.
8. In the first consumer's terminal, press enter.
9. As the first consumer acks messages, observe that only then does the second consumer get any messages.

**Bug scenario 2**
1. Create a **non partitioned** topic.
2. Create a KeyShared subscription on that topic.
3. Produce 100 messages on that topic, each with a distinct partition key (e.g. `key1`, `key2` through `key100`).
4. Start a consumer on the topic with the below code and ensure it prints `Press a key to acknowledge messages`.
5. Start a second consumer on the topic with the below code.
6. Observe that the second consumer does not receive any messages (i.e. it does not print `Press a key to acknowledge messages`), even though [hash range redistribution](https://medium.com/@ankushkhanna1988/apache-pulsar-key-shared-mode-sticky-consistent-hashing-a4ee7133930a) should have allocated at least some of the 100 keys to the new consumer.
8. In the first consumer's terminal, press enter.
9. As the first consumer acks messages, observe that only then does the second consumer get any messages.

**Consumer code**

```python
import time
from pulsar import Client, ConsumerType, Timeout
import os

TOPIC = 'THETOPIC'
SUBSCRIPTION = 'MYSUBSCRIPTION'

def recv(sub):
    while True:
        try:
            msg = sub.receive(100)
            print("Got message", msg.data())
            return msg
        except Timeout:
            pass

def main():
    client = Client(service_url='pulsar://localhost:6650')
    sub = client.subscribe(
        topic=TOPIC,
        subscription_name=SUBSCRIPTION,
        consumer_type=ConsumerType.KeyShared,
        receiver_queue_size=2,
        consumer_name=f'testconsumer-{os.getpid()}'
    )
    msg = recv(sub)
    input("Press a key to acknowledge messages")
    while True:
        sub.acknowledge(msg)
        msg = recv(sub)


if __name__ == '__main__':
    main()
```

**Expected behavior**
In scenario 1, the second consumer should receive at least some messages in step 7.
In scenario 2, the second consumer should receive messages as soon as it starts in step 5.

In short, **I think hash range redistribution is not working right, or is not triggering message re-routing:** when new KeyShared consumers arrive, two things should happen: 
1. New consumers should be allocated part of the hash range of their subscription.
2. Any backlog messages for keys in that range should be sent to the new consumer.

In scenario 1, neither part is working. 
In scenario 2, part 1 is working, but I think part 2 is not.

**Environment:**

Same environment as https://github.com/apache/pulsar/issues/15701

### PR DESCRIPTION
**Pull request by momo-jun**
Fixes #15705 

### Modifications

1. Add a limitation for the `key_shared` subscription type.
2. Apply the new markdown syntax to more notes and tips.

The preview looks good.

image

### Documentation

- [x] `doc`

### CODE REVIEW

**Comment by momo-jun** [comment]
//cc @codelipenghui @Anonymitaet

File: site2/docs/concepts-messaging.md [hunk 1]
```diff
@@ -916,7 +937,9 @@ The other available approach to message deduplication is to ensure that each mes
 
 Message deduplication makes Pulsar an ideal messaging system to be used in conjunction with stream processing engines (SPEs) and other systems seeking to provide effectively-once processing semantics. Messaging systems that do not offer automatic message deduplication require the SPE or other system to guarantee deduplication, which means that strict message ordering comes at the cost of burdening the application with the responsibility of deduplication. With Pulsar, strict ordering guarantees come at no application-level cost.
 
-> You can find more in-depth information in [this post](https://www.splunk.com/en_us/blog/it/exactly-once-is-not-exactly-the-same.html).
+:::tip
+You can find more in-depth information in [this post](https://www.splunk.com/en_us/blog/it/exactly-once-is-not-exactly-the-same.html).
```
Comment (Anonymitaet) [inline site2/docs/concepts-messaging.md:941]:
```suggestion

```
Suggest removing this. 
To keep neutral and not to be challenged by ASF, we avoid adding blogs of commercial companies to the Pulsar documentation.

Comment (momo-jun) [reply site2/docs/concepts-messaging.md]: OK. Removed.

**Comment by momo-jun** [comment]
Hi @codelipenghui can you pls review this PR, and shall we apply this change to 2.10 or even more versions?

File: site2/docs/concepts-messaging.md [hunk 2]
```diff
@@ -633,10 +641,14 @@ producer = client.create_producer(topic='my-topic', batching_type=pulsar.Batchin
...
-> When you use Key_Shared type, be aware that:
-> * You need to specify a key or orderingKey for messages.
-> * You cannot use cumulative acknowledgment with Key_Shared type.
+:::caution
+**Limitations of Key_Shared type** 
+
+When you use Key_Shared type, be aware that:
+  * You need to specify a key or orderingKey for messages.
+  * You cannot use cumulative acknowledgment with Key_Shared type.
+  * When the position of the newest message in a topic is `X`, a key-shared consumer that is newly attached to the same subscription and connected to the topic will **not** receive any messages until the position of the oldest unacknowledged message in this subscription is newer than or equal to `X`.
```
Comment (codelipenghui) [inline site2/docs/concepts-messaging.md:650]:
```suggestion
  * When the position of the newest message in a topic is `X`, a key-shared consumer that is newly attached to the same subscription and connected to the topic will **not** receive any messages until all the messages before `X` have been acknowledged. 
```

**Comment by momo-jun** [comment]
@codelipenghui thanks for your review.  I've incorporated your suggestion. Please take another look.

**Review by codelipenghui — APPROVED** [review]

**Pull request merged**